Remove native_functions.yaml dependency from ScanKernels.cu

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #67216
* #66980
* #66979
* #66914
* #66913
* #66794
* #66793
* #66621
* __->__ #66620

This splits the Tensor-dependant code out into a cpp file.

A slight complicating factor is `scan_dim` using `copy_` to handle
non-contiguous out arguments. So, I've moved that code into the
caller which does introduce some duplication. Though it's only ~10
lines extra in total.

Differential Revision: [D31856106](https://our.internmc.facebook.com/intern/diff/D31856106)

peterbell10 committed:
Remove native_functions.yaml dependency from ScanKernels.cu

This splits the Tensor-dependant code out into a cpp file.

A slight complicating factor is `scan_dim` using `copy_` to handle
non-contiguous out arguments. So, I've moved that code into the
caller which does introduce some duplication. Though it's only ~10
lines extra in total.

[ghstack-poisoned]

peterbell10 committed:
Update on "Remove native_functions.yaml dependency from ScanKernels.cu"


This splits the Tensor-dependant code out into a cpp file.

A slight complicating factor is `scan_dim` using `copy_` to handle
non-contiguous out arguments. So, I've moved that code into the
caller which does introduce some duplication. Though it's only ~10
lines extra in total.

[ghstack-poisoned]

peterbell10 committed:
Update on "Remove native_functions.yaml dependency from ScanKernels.cu"


This splits the Tensor-dependant code out into a cpp file.

A slight complicating factor is `scan_dim` using `copy_` to handle
non-contiguous out arguments. So, I've moved that code into the
caller which does introduce some duplication. Though it's only ~10
lines extra in total.

[ghstack-poisoned]

peterbell10 commented:
@dagitses

peterbell10 committed:
Update on "Remove native_functions.yaml dependency from ScanKernels.cu"


This splits the Tensor-dependant code out into a cpp file.

A slight complicating factor is `scan_dim` using `copy_` to handle
non-contiguous out arguments. So, I've moved that code into the
caller which does introduce some duplication. Though it's only ~10
lines extra in total.

[ghstack-poisoned]

dagitses reviewed (approved):
This is a great approach but I am concerned about how to communicate to other contributors that this ought to be done going forward and how to do it.

The philosophy definitely needs to be documented, but the harder part is making sure everyone knows it or it is discoverable at the right time. Do you have any better ideas than linking it extensively in CUDA implementation files?

For example, I could see something like:
`#define TORCH_ASSERT_NO_OPERATORS  // see link to document`

peterbell10 committed:
Update on "Remove native_functions.yaml dependency from ScanKernels.cu"


This splits the Tensor-dependant code out into a cpp file.

A slight complicating factor is `scan_dim` using `copy_` to handle
non-contiguous out arguments. So, I've moved that code into the
caller which does introduce some duplication. Though it's only ~10
lines extra in total.

Differential Revision: [D31856106](https://our.internmc.facebook.com/intern/diff/D31856106)

[ghstack-poisoned]

Commit:
Update on "Remove native_functions.yaml dependency from ScanKernels.cu"


This splits the Tensor-dependant code out into a cpp file.

A slight complicating factor is `scan_dim` using `copy_` to handle
non-contiguous out arguments. So, I've moved that code into the
caller which does introduce some duplication. Though it's only ~10
lines extra in total.

Differential Revision: [D31856106](https://our.internmc.facebook.com/intern/diff/D31856106)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31856106).
